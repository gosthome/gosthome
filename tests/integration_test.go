package tests_test

import (
	"bytes"
	"context"
	"flag"
	"log"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"testing"
	"text/template"

	"github.com/gkampitakis/go-snaps/snaps"
	_ "github.com/gosthome/gosthome/components"
	"github.com/gosthome/gosthome/components/api/esphomeproto"
	"github.com/gosthome/gosthome/components/api/frameshakers"
	"github.com/gosthome/gosthome/core"
	"github.com/gosthome/gosthome/core/config"
	"github.com/gosthome/gosthome/tests"
)

var sampleConfig = template.Must(template.New("").Parse(`
gosthome:
    name: testABC
    friendly_name: "Testing ABC"
    mac: {{ .MAC }}

api:
    address: "127.0.0.1"
    port: {{ .Port }}
    {{ if ne .Password ""}}password: "{{ .Password }}"{{end }}
    {{ if ne .NoisePSK ""}}
    encryption:
        key: "{{.NoisePSK }}"
    {{ end }}

demo:
`))

var debug = flag.Bool("debug", false, "debug python output")

func TestGoServerPyClient(t *testing.T) {
	if testing.Verbose() {
		slog.SetDefault(slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
			Level: slog.LevelDebug,
		})))
	}
	password := "xxxLongandUniquepasswordStringxxx"
	noise, err := frameshakers.GenerateEncryptionKey()
	if err != nil {
		t.Fatal(err)
	}
	nodeMac, err := config.GenerateMAC()
	if err != nil {
		t.Fatal(err)
	}

	type tcase struct {
		name     string
		password string
		noise    *frameshakers.ConfigNoisePSK
	}

	cases := []tcase{
		{
			name:     "TestPassword",
			password: password,
			noise:    nil,
		},
		{
			name:     "TestEncryption",
			password: "",
			noise:    noise,
		},
		{
			name:     "TestEncryptionAndPassword",
			password: password,
			noise:    noise,
		},
	}

	for _, tcase := range cases {
		t.Run(tcase.name, func(t *testing.T) {
			t.Parallel()
			esphomeprotoPort := tests.GetFreePort(t)
			configBytes := &bytes.Buffer{}
			err = sampleConfig.Execute(configBytes, &struct {
				Port     int
				Password string
				NoisePSK string
				MAC      string
			}{
				Port:     esphomeprotoPort,
				Password: tcase.password,
				NoisePSK: tcase.noise.String(),
				MAC:      nodeMac.String(),
			})
			if err != nil {
				t.Fatal(err)
			}
			configData := make([]byte, configBytes.Len())
			copy(configData, configBytes.Bytes())
			cfg, err := config.LoadConfig(configBytes)
			if err != nil {
				print(string(configData))
				t.Fatal(err)
			}

			n, err := core.NewNode(context.Background(), cfg)
			if err != nil {
				t.Fatal(err)
			}
			defer func() {
				if err = n.Close(); err != nil {
					t.Error(err)
				}
			}()
			n.Start()

			testdata := filepath.Join("..", "components", "api", "esphomeproto", "testdata")

			python := filepath.Join(testdata, "setup.sh")
			// For debugging. Don't key on testing.Verbose() since the test would be
			// failing.
			if *debug {
				cmd := exec.Command(
					python, filepath.Join(testdata, "test.py"),
					"--port", strconv.Itoa(esphomeprotoPort),
					"--password", tcase.password,
					"--noise-psk", tcase.noise.String(),
					"--verbose")
				cmd.Stdout = os.Stdout
				cmd.Stderr = os.Stderr
				if err = cmd.Run(); err != nil {
					t.Error(err)
				}
				t.FailNow()
			}
			out, err := exec.Command(
				python, filepath.Join(testdata, "test.py"),
				"--port", strconv.Itoa(esphomeprotoPort),
				"--password", tcase.password,
				"--noise-psk", tcase.noise.String(),
			).CombinedOutput()
			if err != nil {
				t.Error(err)
			}
			got := string(out)
			if runtime.GOOS == "windows" {
				got = strings.ReplaceAll(got, "\r", "")
			}
			if tcase.password != "" {
				got = strings.ReplaceAll(got, tcase.password, "<password_redacted>")
			}
			got = strings.ReplaceAll(got, runtime.GOOS, "<goos>")
			got = strings.ReplaceAll(got, runtime.GOARCH, "<goarch>")
			got = strings.ReplaceAll(got, cfg.Gosthome.MAC.String(), "<mac>")
			got = strings.ReplaceAll(got, core.Version(), "<gosthome_version>")
			got = strings.ReplaceAll(got, esphomeproto.ESPHOME_VERSION, "<esphome_version>")

			snaps.MatchSnapshot(t, got)
		})
	}
}

var sampleEspHostConfig = template.Must(template.New("").Parse(`
esphome:
    name: test-abc{{ if ne .Password ""}}-pw{{end }}{{ if ne .NoisePSK ""}}-noise{{end }}
    friendly_name: "Testing ABC"

host:
    mac: {{ .MAC }}

api:
    address: "127.0.0.1"
    port: {{ .Port }}
    {{ if ne .Password ""}}password: "{{ .Password }}"{{end }}
    {{ if ne .NoisePSK ""}}
    encryption:
        key: "{{.NoisePSK }}"
    {{ end }}

demo:
`))

// func TestGoClientPyServer(t *testing.T) {
// 	if testing.Verbose() {
// 		slog.SetDefault(slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
// 			Level: slog.LevelDebug,
// 		})))
// 	}
// 	password := "abc"
// 	noise, err := frameshakers.GenerateEncryptionKey()
// 	if err != nil {
// 		t.Fatal(err)
// 	}
// 	nodeMac, err := config.GenerateMAC()
// 	if err != nil {
// 		t.Fatal(err)
// 	}

// 	type tcase struct {
// 		name     string
// 		password string
// 		noise    *frameshakers.ConfigNoisePSK
// 	}

// 	cases := []tcase{
// 		{
// 			name:     "TestPassword",
// 			password: password,
// 			noise:    nil,
// 		},
// 		{
// 			name:     "TestEncryption",
// 			password: "",
// 			noise:    noise,
// 		},
// 		{
// 			name:     "TestEncryptionAndPassword",
// 			password: password,
// 			noise:    noise,
// 		},
// 	}

// 	for _, tcase := range cases {
// 		t.Run(tcase.name, func(t *testing.T) {
// 			t.Parallel()
// 			esphomeprotoPort := tests.GetFreePort(t)
// 			configBytes := &bytes.Buffer{}
// 			err = sampleConfig.Execute(configBytes, &struct {
// 				Port     int
// 				Password string
// 				NoisePSK string
// 				MAC      string
// 			}{
// 				Port:     esphomeprotoPort,
// 				Password: tcase.password,
// 				NoisePSK: tcase.noise.String(),
// 				MAC:      nodeMac.String(),
// 			})
// 			if err != nil {
// 				t.Fatal(err)
// 			}
// 			configData := make([]byte, configBytes.Len())
// 			copy(configData, configBytes.Bytes())
// 			cfg, err := config.LoadConfig(configBytes)
// 			if err != nil {
// 				print(string(configData))
// 				t.Fatal(err)
// 			}

// 			n, err := core.NewNode(context.Background(), cfg)
// 			if err != nil {
// 				t.Fatal(err)
// 			}
// 			defer func() {
// 				if err = n.Close(); err != nil {
// 					t.Error(err)
// 				}
// 			}()
// 			n.Start()

// 			testdata := filepath.Join("..", "components", "api", "esphomeproto", "testdata")

// 			python := filepath.Join(testdata, "setup.sh")
// 			// For debugging. Don't key on testing.Verbose() since the test would be
// 			// failing.
// 			if *debug {
// 				cmd := exec.Command(
// 					python, filepath.Join(testdata, "test.py"),
// 					"--port", strconv.Itoa(esphomeprotoPort),
// 					"--password", tcase.password,
// 					"--noise-psk", tcase.noise.String(),
// 					"--verbose")
// 				cmd.Stdout = os.Stdout
// 				cmd.Stderr = os.Stderr
// 				if err = cmd.Run(); err != nil {
// 					t.Error(err)
// 				}
// 				t.FailNow()
// 			}
// 			out, err := exec.Command(
// 				python, filepath.Join(testdata, "test.py"),
// 				"--port", strconv.Itoa(esphomeprotoPort),
// 				"--password", tcase.password,
// 				"--noise-psk", tcase.noise.String(),
// 			).CombinedOutput()
// 			if err != nil {
// 				t.Error(err)
// 			}
// 			got := string(out)
// 			if runtime.GOOS == "windows" {
// 				got = strings.ReplaceAll(got, "\r", "")
// 			}

// 		})
// 	}
// }

func init() {
	log.SetFlags(log.Ldate | log.Ltime | log.Lmicroseconds | log.Lshortfile)
}
