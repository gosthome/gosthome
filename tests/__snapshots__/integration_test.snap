
[TestGoServerPyClient/TestEncryption - 1]
API version: APIVersion(major=1, minor=10)
Device info: DeviceInfo(uses_password=False, name='testABC', friendly_name='Testing ABC', mac_address='<mac>', compilation_time='2022', model='<goos>/<goarch>', manufacturer='gosthome', has_deep_sleep=False, esphome_version='<esphome_version>', project_name='', project_version='', webserver_port=0, legacy_voice_assistant_version=0, voice_assistant_feature_flags=0, legacy_bluetooth_proxy_version=0, bluetooth_proxy_feature_flags=0, suggested_area='')

Entities:
- BinarySensorInfo(object_id='demo_movement_backyard', key=1756138606, name='Demo Movement Backyard', unique_id='testABCbinary_sensordemo_movement_backyard', disabled_by_default=False, icon='', entity_category=<EntityCategory.NONE: 0>, <gosthome_version>ice_class='motion', is_status_binary_sensor=False)
- BinarySensorInfo(object_id='demo_basement_floor_wet', key=2292024046, name='Demo Basement Floor Wet', unique_id='testABCbinary_sensordemo_basement_floor_wet', disabled_by_default=False, icon='', entity_category=<EntityCategory.NONE: 0>, <gosthome_version>ice_class='moisture', is_status_binary_sensor=False)
- ButtonInfo(object_id='demo_regenerate_seed', key=258008683, name='Demo Regenerate Seed', unique_id='testABCbuttondemo_regenerate_seed', disabled_by_default=False, icon='', entity_category=<EntityCategory.CONFIG: 1>, <gosthome_version>ice_class='restart')
- ButtonInfo(object_id='demo_regenerate_seed_config', key=1323890508, name='Demo Regenerate Seed Config', unique_id='testABCbuttondemo_regenerate_seed_config', disabled_by_default=False, icon='', entity_category=<EntityCategory.CONFIG: 1>, <gosthome_version>ice_class='restart')

State:
- BinarySensorState(key=1756138606, state=False, missing_state=False)
- BinarySensorState(key=2292024046, state=False, missing_state=False)

---

[TestGoServerPyClient/TestEncryptionAndPassword - 1]
API version: APIVersion(major=1, minor=10)
Device info: DeviceInfo(uses_password=True, name='testABC', friendly_name='Testing ABC', mac_address='<mac>', compilation_time='2022', model='<goos>/<goarch>', manufacturer='gosthome', has_deep_sleep=False, esphome_version='<esphome_version>', project_name='', project_version='', webserver_port=0, legacy_voice_assistant_version=0, voice_assistant_feature_flags=0, legacy_bluetooth_proxy_version=0, bluetooth_proxy_feature_flags=0, suggested_area='')

Entities:
- BinarySensorInfo(object_id='demo_movement_backyard', key=1756138606, name='Demo Movement Backyard', unique_id='testABCbinary_sensordemo_movement_backyard', disabled_by_default=False, icon='', entity_category=<EntityCategory.NONE: 0>, <gosthome_version>ice_class='motion', is_status_binary_sensor=False)
- BinarySensorInfo(object_id='demo_basement_floor_wet', key=2292024046, name='Demo Basement Floor Wet', unique_id='testABCbinary_sensordemo_basement_floor_wet', disabled_by_default=False, icon='', entity_category=<EntityCategory.NONE: 0>, <gosthome_version>ice_class='moisture', is_status_binary_sensor=False)
- ButtonInfo(object_id='demo_regenerate_seed', key=258008683, name='Demo Regenerate Seed', unique_id='testABCbuttondemo_regenerate_seed', disabled_by_default=False, icon='', entity_category=<EntityCategory.CONFIG: 1>, <gosthome_version>ice_class='restart')
- ButtonInfo(object_id='demo_regenerate_seed_config', key=1323890508, name='Demo Regenerate Seed Config', unique_id='testABCbuttondemo_regenerate_seed_config', disabled_by_default=False, icon='', entity_category=<EntityCategory.CONFIG: 1>, <gosthome_version>ice_class='restart')

State:
- BinarySensorState(key=1756138606, state=False, missing_state=False)
- BinarySensorState(key=2292024046, state=False, missing_state=False)

---

[TestGoServerPyClient/TestPassword - 1]
API version: APIVersion(major=1, minor=10)
Device info: DeviceInfo(uses_password=True, name='testABC', friendly_name='Testing ABC', mac_address='<mac>', compilation_time='2022', model='<goos>/<goarch>', manufacturer='gosthome', has_deep_sleep=False, esphome_version='<esphome_version>', project_name='', project_version='', webserver_port=0, legacy_voice_assistant_version=0, voice_assistant_feature_flags=0, legacy_bluetooth_proxy_version=0, bluetooth_proxy_feature_flags=0, suggested_area='')

Entities:
- BinarySensorInfo(object_id='demo_movement_backyard', key=1756138606, name='Demo Movement Backyard', unique_id='testABCbinary_sensordemo_movement_backyard', disabled_by_default=False, icon='', entity_category=<EntityCategory.NONE: 0>, <gosthome_version>ice_class='motion', is_status_binary_sensor=False)
- BinarySensorInfo(object_id='demo_basement_floor_wet', key=2292024046, name='Demo Basement Floor Wet', unique_id='testABCbinary_sensordemo_basement_floor_wet', disabled_by_default=False, icon='', entity_category=<EntityCategory.NONE: 0>, <gosthome_version>ice_class='moisture', is_status_binary_sensor=False)
- ButtonInfo(object_id='demo_regenerate_seed', key=258008683, name='Demo Regenerate Seed', unique_id='testABCbuttondemo_regenerate_seed', disabled_by_default=False, icon='', entity_category=<EntityCategory.CONFIG: 1>, <gosthome_version>ice_class='restart')
- ButtonInfo(object_id='demo_regenerate_seed_config', key=1323890508, name='Demo Regenerate Seed Config', unique_id='testABCbuttondemo_regenerate_seed_config', disabled_by_default=False, icon='', entity_category=<EntityCategory.CONFIG: 1>, <gosthome_version>ice_class='restart')

State:
- BinarySensorState(key=1756138606, state=False, missing_state=False)
- BinarySensorState(key=2292024046, state=False, missing_state=False)

---
